model-optimizer/mo/ops/slice.py

```diff
@@ -13,133 +13,178 @@
+        if starts is None or ends is None:
+            raise Error('The non-constant start/end values for Slice operation "{}" are not supported'.format(node.name))
 
-        # If value is None create dummy value for shape propagation
-        if value is None:
-            value = np.zeros(input_shape)
-
-        # Following ONNX and TF specification, in case of unknown axis, axises should be in greater order
-        if axis is None:
-            axis = [x for x in range(len(start))]
+        if node.is_in_port_connected(3):
+            axes = node.in_port(3).data.get_value()
+            if axes is None:
+                raise Error('The non-constant axes values for Slice operation "{}" is not supported'.format(node.name))
+        else:
+            axes = [x for x in range(len(starts))]
 
-        if steps is None:
-            steps = np.ones(start.size, dtype=np.int64)
+        if node.is_in_port_connected(4):
+            steps = node.in_port(4).data.get_value()
+            if steps is None:
+                raise Error('The non-constant steps values for Slice operation "{}" is not supported'.format(node.name))
+        else:
+            steps = np.ones(len(starts), dtype=np.int64)
 
-        # Calculate output value for slice operation
-        slice_idx = [None for x in range(len(node.in_node().shape))]
-        shrink_axis_mask = [False for x in range(len(node.in_node().shape))]
-        for id in range(len(axis)):
+        slice_idx = [slice(0, in_shape, 1) for in_shape in input_shape]
+        for i in range(len(axes)):
             # Ranged for output value for specified axis
-            slice_idx[axis[id]] = slice(start[id], end[id], steps[id])
+            slice_idx[axes[i]] = slice(starts[i], ends[i], steps[i])
 
-        for axis, s in enumerate(slice_idx):
-            if s is None:
-                slice_idx[axis] = slice(0, input_shape[axis], 1)
-
-        # Add new parameters to node
-        node['slices'] = np.array(slice_idx)
-        node['shrink_axis_mask'] = np.array(shrink_axis_mask)
-
-        value = value[tuple(slice_idx)]
-        node.out_node().value = value.copy() if node.in_node(0).value is not None else None
-        node.out_node().shape = np.array(value.shape)
+        if input_value is None:
+            output_shape = get_shape_after_slice(input_shape, slice_idx)
+            node.out_port(0).data.set_shape(output_shape)
+        else:
+            node.out_port(0).data.set_value(input_value[tuple(slice_idx)])
+
+
+def get_shape_after_slice(input_shape: np.ndarray, slice_idx: List) -> np.ndarray:
+    """
+    Calculate shape of a tensor after slicing without actually creating the resulting tensor.
+    Is introduced to save memory.
+    """
+    output_shape = np.zeros(len(input_shape), dtype=np.int32)
+    for i, s in enumerate(slice_idx):
+        start, end = normalize_slice_indices(input_shape[i], s.start, s.stop)
+        output_shape[i] = (end - start) / s.step
```

Comment:
Some rounding should be applied here. Also, should we check that the result of the division is an integer number (not 4.75, for example)?

Reply:
The casting is done automatically because `output_shape` has int dtype. But your question prompted me to check cases when they are not exactly divisible and for that cases my solution didn't work always properly. Improved it. Now it returns always results same as numpy. Please review.